src/net/server/bytearrstream.rs

```diff
@@ -0,0 +1,452 @@
+            self.buffered_size -= self.size_in_bytes;
+            Some(out_bytes)
+        } else {
+            // First buffer is too small, so we need to take across buffers.
+            // Will require a memcopy of u8's across buffers in order to create 1 contiguous buffer
+            // to return.
+            // Removes expended buffers after taking
+
+            let mut out_bytes = Vec::with_capacity(self.size_in_bytes as usize);
+            // this must loop through the bytes buffers because we don't know how many buffers will
+            // be needed to fulfill `size_in_bytes`. e.g. if every buffer had length 1, we'd need to
+            // visit `size_in_bytes` buffers in order to fill `out_bytes`
+            loop {
+                let remaining_bytes = out_bytes.capacity() - out_bytes.len();
+                if remaining_bytes == 0 {
+                    break;
+                }
+                // current `buffer` has more bytes than needed
+                if self.buffered[0].len() > remaining_bytes {
+                    let remaining = self.buffered[0].split_to(remaining_bytes);
+                    out_bytes.extend_from_slice(&remaining);
+                } else {
+                    // current `buffer` has <= bytes needed, remove and append to out_bytes
+                    out_bytes.extend_from_slice(&self.buffered.pop_front().unwrap());
+                }
+            }
+            // reduce size of total buffers accordingly
+            self.buffered_size -= self.size_in_bytes;
+            Some(Bytes::from(out_bytes))
+        }
+    }
+
+    fn push_back(&mut self, buf: Bytes) {
+        self.buffered_size += u32::try_from(buf.len()).unwrap();
+        self.buffered.push_back(buf);
+    }
+}
+
+impl Stream for ByteArrStream {
+    type Item = Result<Bytes, std::io::Error>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            // if we have enough bytes buffered, return them
+            if let Some(bytes) = self.pop_front() {
+                return Poll::Ready(Some(Ok(bytes)));
+            }
+            // if we need more bytes, poll the body
+            let mut this = self.as_mut().project();
+            match ready!(this.body.as_mut().poll_next(cx)) {
+                // if body is expended, but we have some bytes leftover, error
+                None if *this.buffered_size > 0 => {
+                    return Poll::Ready(Some(Err(std::io::Error::new::<BoxError>(
+                        std::io::ErrorKind::WriteZero,
+                        format!(
+                            "expected body to align on size {}, but has insufficient bytes {}",
+                            *this.size_in_bytes, *this.buffered_size
+                        )
+                        .into(),
+                    ))));
+                }
+                // if body is finished, this stream is finished
+                None => return Poll::Ready(None),
```

Comment:
This should produce an error if the stream ends when the buffered_size is not zero.  In that case there will be a partial value present.

Reply:
the case above this one is

```rust
None if *this.buffered_size > 0 => {
```
is that sufficient?

Reply:
we could add `None if this.buffered_size == 0` to make it more clear but yea I got tripped here too. this code I believe handles the case when stream ends producing insufficient bytes to create another chunk.

We would likely want to change that though and use two parameters for this stream `alignment` and `buffer_size`: https://github.com/private-attribution/ipa/pull/344/files#r1048124439